subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptor.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath.intercept;
+
+import org.codehaus.groovy.runtime.callsite.AbstractCallSite;
+import org.codehaus.groovy.runtime.callsite.CallSite;
+
+/**
+ * Intercepts method and constructor calls as well as property reads in dynamic Groovy bytecode.
+ * <p>
+ * The interceptor serves as a factory for decorated Groovy {@link CallSite} instances.
+ * These instances replace original CallSites where the calls are to be intercepted.
+ * <p>
+ * Descendants of this class should be thread-safe, making a stateless implementation is perfect.
+ */
+public abstract class CallInterceptor {
+    /**
+     * Represents a single invocation of the intercepted method/constructor/property.
+     */
+    public interface Invocation {
+        /**
+         * Returns the receiver of the invocation.
+         * It can be a {@link Class} if the invocation targets constructor, static method, or static property.
+         * It can be the instance if the invocation targets the instance method or property.
+         *
+         * @return the receiver of the method
+         * @see CallSite
+         */
+        Object getReceiver();
+
+        /**
+         * Returns a number of arguments supplied for this invocation.
+         */
+        int getArgsCount();
+
+        /**
+         * Returns an <b>unwrapped</b> argument at the position {@code pos}.
+         * Arguments are numbered left-to-right, from 0 to {@code getArgsCount() - 1} inclusive.
+         * Throws {@link ArrayIndexOutOfBoundsException} if {@code pos} is outside the bounds.
+         *
+         * @param pos the position of the argument
+         * @return the unwrapped value of the argument
+         */
+        Object getArgument(int pos);
+
+        /**
+         * Returns an <b>unwrapped</b> argument at the position {@code pos} or {@code null} if the {@code pos} is greater or equal than {@link #getArgsCount()}.
+         * This method is useful for handling "telescopic" overloads.
```

Comment:
🗒️ An example of a "telescopic" overload would be useful here.

Reply:
Done